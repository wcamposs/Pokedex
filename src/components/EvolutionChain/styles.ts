import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
    container: {
        flex: 1,
        marginVertical: 8,
        marginHorizontal: 12,
    },
    row: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    evolutionInnerContainer: (backgroundColor) => ({
        backgroundColor,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 30,
    }),
    evolutionRow: {
        flexDirection: 'row',
        alignItems: 'center',
        marginBottom: 12,
    },
    // image container
    evolutionImageContainer: {
        justifyContent: 'center',
        alignItems: 'center',
        marginBottom: 16,
    },
    evolutionPokemonTitle: {
        marginTop: 20,
        marginBottom: 28,
        textAlign: 'center',
        fontSize: 16,
        fontWeight: '700',
        textTransform: 'capitalize',
    },
    imageContainer: {
        marginBottom: 28,
        marginHorizontal: 16,
    },
    evolutionImage: {
        margin: -40,
        width: 150,
        height: 150,
    },
    separatorContainer: {
        alignItems: 'center',
        justifyContent: 'center',
        marginHorizontal: 12,
    },
    evolutionLevelText: {
        marginTop: 20,
        marginBottom: 8,
        textAlign: 'center',
        fontSize: 14,
        fontWeight: '700',
    },
    divider: {
        marginHorizontal: 8,
        marginBottom: 10,
        width: 50,
        height: 1,
        borderWidth: 2,
    },
    emptyContainer: {
        marginVertical: 20,
    },
});

export default styles;
