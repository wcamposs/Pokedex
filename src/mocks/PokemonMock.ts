const PokemonMock = {
	name: 'ivysaur',
	url: 'https://pokeapi.co/api/v2/pokemon/2/',
	id: '2',
	spriteUrl:
		'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
	types: [
		{
			slot: 1,
			type: {
				name: 'grass',
				url: 'https://pokeapi.co/api/v2/type/12/',
			},
		},
		{
			slot: 2,
			type: {
				name: 'poison',
				url: 'https://pokeapi.co/api/v2/type/4/',
			},
		},
	],
	abilities: [
		{
			ability: {
				name: 'overgrow',
				url: 'https://pokeapi.co/api/v2/ability/65/',
			},
			is_hidden: false,
			slot: 1,
		},
		{
			ability: {
				name: 'chlorophyll',
				url: 'https://pokeapi.co/api/v2/ability/34/',
			},
			is_hidden: true,
			slot: 3,
		},
	],
	height: 10,
	weight: 130,
	stats: [
		{
			base_stat: 60,
			effort: 0,
			stat: {
				name: 'hp',
				url: 'https://pokeapi.co/api/v2/stat/1/',
			},
		},
		{
			base_stat: 62,
			effort: 0,
			stat: {
				name: 'attack',
				url: 'https://pokeapi.co/api/v2/stat/2/',
			},
		},
		{
			base_stat: 63,
			effort: 0,
			stat: {
				name: 'defense',
				url: 'https://pokeapi.co/api/v2/stat/3/',
			},
		},
		{
			base_stat: 80,
			effort: 1,
			stat: {
				name: 'special-attack',
				url: 'https://pokeapi.co/api/v2/stat/4/',
			},
		},
		{
			base_stat: 80,
			effort: 1,
			stat: {
				name: 'special-defense',
				url: 'https://pokeapi.co/api/v2/stat/5/',
			},
		},
		{
			base_stat: 60,
			effort: 0,
			stat: {
				name: 'speed',
				url: 'https://pokeapi.co/api/v2/stat/6/',
			},
		},
	],
	evolutionChain: {
		evolution_details: [],
		evolves_to: [
			{
				evolution_details: [
					{
						gender: null,
						held_item: null,
						item: null,
						known_move: null,
						known_move_type: null,
						location: null,
						min_affection: null,
						min_beauty: null,
						min_happiness: null,
						min_level: 16,
						needs_overworld_rain: false,
						party_species: null,
						party_type: null,
						relative_physical_stats: null,
						time_of_day: '',
						trade_species: null,
						trigger: {
							name: 'level-up',
							url: 'https://pokeapi.co/api/v2/evolution-trigger/1/',
						},
						turn_upside_down: false,
					},
				],
				evolves_to: [
					{
						evolution_details: [
							{
								gender: null,
								held_item: null,
								item: null,
								known_move: null,
								known_move_type: null,
								location: null,
								min_affection: null,
								min_beauty: null,
								min_happiness: null,
								min_level: 32,
								needs_overworld_rain: false,
								party_species: null,
								party_type: null,
								relative_physical_stats: null,
								time_of_day: '',
								trade_species: null,
								trigger: {
									name: 'level-up',
									url: 'https://pokeapi.co/api/v2/evolution-trigger/1/',
								},
								turn_upside_down: false,
							},
						],
						evolves_to: [],
						is_baby: false,
						species: {
							name: 'venusaur',
							url: 'https://pokeapi.co/api/v2/pokemon-species/3/',
							spriteUrl:
								'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png',
						},
					},
				],
				is_baby: false,
				species: {
					name: 'ivysaur',
					url: 'https://pokeapi.co/api/v2/pokemon-species/2/',
					spriteUrl:
						'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
				},
			},
		],
		is_baby: false,
		species: {
			name: 'bulbasaur',
			url: 'https://pokeapi.co/api/v2/pokemon-species/1/',
			spriteUrl:
				'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
		},
	},
	growthRate: {
		name: 'medium-slow',
		url: 'https://pokeapi.co/api/v2/growth-rate/4/',
	},
	habitat: {
		name: 'grassland',
		url: 'https://pokeapi.co/api/v2/pokemon-habitat/3/',
	},
	legendary: false,
	mythical: false,
};

export default PokemonMock;
