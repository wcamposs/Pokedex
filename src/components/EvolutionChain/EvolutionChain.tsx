// libraries
import React from 'react';
import { Image, Text, View } from 'react-native';
import { Divider } from 'react-native-elements';

// js
import styles from './styles';

function EvolutionChain(props) {
	const { backgroundColor, evolutionChain = [] } = props;
	const { species, evolves_to = [] } = evolutionChain;

	const { evolves_to: secondEvolutionChain = [] } = evolutionChain || [];
	const { species: secondGenSpecies } = secondEvolutionChain[0] || {};

	const { evolves_to: thirdEvolutionChain = [] } =
		secondEvolutionChain[0] || {};
	const { species: thirdGenSpecies } = thirdEvolutionChain[0] || {};

	const { evolves_to: fourthEvolutionChain = [] } =
		thirdEvolutionChain[0] || {};

	function renderEvolutionContainer(name: string, imageUrl: string) {
		return (
			<View style={styles.evolutionImageContainer}>
				<Text style={styles.evolutionPokemonTitle}>{name}</Text>
				<View style={styles.imageContainer}>
					<Image
						source={{ uri: imageUrl }}
						style={styles.evolutionImage}
						resizeMode="contain"
					/>
				</View>
			</View>
		);
	}

	function renderSeparatorContainer(
		hasMinLevel: boolean,
		pokemonLevel: number,
	) {
		return (
			<View style={styles.separatorContainer}>
				{!hasMinLevel ? (
					<View style={styles.emptyContainer} />
				) : (
					<Text style={styles.evolutionLevelText}>
						{`LVL: ${pokemonLevel}`}
					</Text>
				)}
				<Divider style={styles.divider} />
			</View>
		);
	}

	return (
		<View style={styles.evolutionInnerContainer(backgroundColor)}>
			{/* first evolution to second */}
			{!evolves_to.length ? (
				<>{renderEvolutionContainer(species.name, species.spriteUrl)}</>
			) : (
				<>
					{secondEvolutionChain.map((evolution: any) => {
						const {
							species: secondGenSpecies = {},
							evolution_details: secondGenEvolutionDetails = [],
						} = evolution || {};

						const hasMinLvl =
							secondGenEvolutionDetails[0]?.min_level;

						return (
							<View style={styles.evolutionRow}>
								{renderEvolutionContainer(
									species.name,
									species.spriteUrl,
								)}

								{renderSeparatorContainer(
									hasMinLvl,
									secondGenEvolutionDetails[0].min_level,
								)}

								{renderEvolutionContainer(
									secondGenSpecies.name,
									secondGenSpecies.spriteUrl,
								)}
							</View>
						);
					})}
				</>
			)}
			{/* second evolution to third */}
			{!thirdEvolutionChain.length ? null : (
				<>
					{thirdEvolutionChain.map((evolution: any) => {
						const {
							species: thirdGenSpecies = {},
							evolution_details: thirdGenEvolutionDetails = [],
						} = evolution || {};

						const hasSecondEvoMinLvl =
							thirdGenEvolutionDetails[0]?.min_level;

						return (
							<View style={styles.evolutionRow}>
								{renderEvolutionContainer(
									secondGenSpecies.name,
									secondGenSpecies.spriteUrl,
								)}

								{renderSeparatorContainer(
									hasSecondEvoMinLvl,
									thirdGenEvolutionDetails[0].min_level,
								)}

								{renderEvolutionContainer(
									thirdGenSpecies.name,
									thirdGenSpecies.spriteUrl,
								)}
							</View>
						);
					})}
				</>
			)}
			{/* third evolution to fourth */}
			{!fourthEvolutionChain.length ? null : (
				<>
					{fourthEvolutionChain.map((evolution: any) => {
						const {
							species: fourthGenSpecies = {},
							evolution_details: fourthGenEvolutionDetails = [],
						} = evolution || {};

						const hasThirdEvoMinLvl =
							fourthGenEvolutionDetails[0]?.min_level;

						return (
							<View style={styles.evolutionRow}>
								{renderEvolutionContainer(
									thirdGenSpecies.name,
									thirdGenSpecies.spriteUrl,
								)}

								{renderSeparatorContainer(
									hasThirdEvoMinLvl,
									fourthGenEvolutionDetails[0].min_level,
								)}

								{renderEvolutionContainer(
									fourthGenSpecies.name,
									fourthGenSpecies.spriteUrl,
								)}
							</View>
						);
					})}
				</>
			)}
		</View>
	);
}

export default EvolutionChain;
