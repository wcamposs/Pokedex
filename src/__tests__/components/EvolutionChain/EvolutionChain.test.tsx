import React from 'react';
import renderer from 'react-test-renderer';

import PokemonMock from '../../../mocks/PokemonMock';

import EvolutionChain from '../../../components/EvolutionChain/EvolutionChain';

jest.useFakeTimers();

describe('<EvolutionChain />', () => {
	it('renders correctly with an evolution chain', () => {
		const mockedEvolutionChain = PokemonMock.evolutionChain;

		const tree = renderer
			.create(
				<EvolutionChain
					backgroundColor="#00bf30"
					evolutionChain={mockedEvolutionChain}
				/>,
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders correctly with an empty evolution chain', () => {
		const tree = renderer
			.create(
				<EvolutionChain
					backgroundColor="#00bf30"
					evolutionChain={[]}
				/>,
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
