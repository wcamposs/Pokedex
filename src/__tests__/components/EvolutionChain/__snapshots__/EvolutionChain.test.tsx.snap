// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EvolutionChain /> renders correctly with an empty evolution chain 1`] = `null`;

exports[`<EvolutionChain /> renders correctly with an evolution chain 1`] = `null`;
